src/main/java/org/openstreetmap/atlas/checks/validation/tag/RoadNameGapCheck.java

```diff
@@ -0,0 +1,210 @@
+    {
+        private Edge edge;
+        private EdgeDirectionComparator edgeDirectionComparator = new EdgeDirectionComparator();
+
+        EdgePredicate(final Edge edge)
+        {
+            this.edge = edge;
+        }
+
+        boolean isSameHeading(final AtlasObject object)
+        {
+            final Edge otherEdge = (Edge) object;
+            return object != null
+                    && this.edgeDirectionComparator.isSameDirection(this.edge, otherEdge, false);
+        }
+    }
+
+    // You can use serialver to regenerate the serial UID.
+    private static final long serialVersionUID = 7104778218412127847L;
+    private static final List<String> VALID_HIGHWAY_TAG_DEFAULT = Arrays.asList("primary",
+            "secondary", "tertiary", "trunk", "motorway");
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays.asList("Edge name is empty.",
+            "Edge name {1} is different from in edge name and out edge name.");
+
+    private final List<String> validHighwayTag;
+
+    /**
+     * The default constructor that must be supplied. The Atlas Checks framework will generate the
+     * checks with this constructor, supplying a configuration that can be used to adjust any
+     * parameters that the check uses during operation.
+     *
+     * @param configuration
+     *            the JSON configuration for this check
+     */
+    public RoadNameGapCheck(final Configuration configuration)
+    {
+        super(configuration);
+        this.validHighwayTag = configurationValue(configuration, "valid.highway.tag",
+                VALID_HIGHWAY_TAG_DEFAULT);
+    }
+
+    /**
+     * This function will validate if the supplied atlas object is valid for the check.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge && Edge.isMasterEdgeIdentifier(object.getIdentifier())
+                && !JunctionTag.isRoundabout(object)
+                && Validators.hasValuesFor(object, HighwayTag.class)
+                && this.validHighwayTag.contains(object.tag(HighwayTag.KEY));
+    }
+
+    /**
+     * This is the actual function that will check to see whether the object needs to be flagged.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        final EdgePredicate edgePredicate = new EdgePredicate(edge);
+        final Set<Edge> inEdges = edge.inEdges().stream().filter(this::validCheckForObject)
+                .filter(edgePredicate::isSameHeading).collect(Collectors.toSet());
```

Comment:
Couldn't we just use a single filter and put validCheckForObject and isSameHeading in the same filter?